### PR TITLE
[WIP] Bug/gh66 verify that iteration setup.is deleted behavior is as per 10 25; fixes #66

### LINKED ISSUE #66: Verify that IterationSetup.IsDeleted behavior is as per 10-25
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the CDP4 Web Services
- [x] I have searched [open](https://github.com/RHEAGROUP/CDP4-WebServices-Community-Edition/issues) and [closed](https://github.com/RHEAGROUP/CDP4-WebServices-Community-Edition/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
Verify that when an IterationSetup is deleted that
  - set the IterationSetup.IsDeleted flag to true and then break out of delete handling (using onBeforeDelete SideEffect)
  - NOT delete the iteration object itself or any other thing
  - throw an exception if invoked on a non-frozen IterationSetup

An IterationSetup that has IsDeleted =true shall:
  - prevent any read of the associated Iteration data (on any level within the Iteration)
  - prevent listing of the Iteration (with corresponding IterationSetup.isDeleted=true) objects when requesting EngineeringModel.iteration listing

### Steps to Reproduce
NA

### PR DESCRIPTION
**Pull request by kamilatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/CDP4-WebServices-Community-Edition/pulls) open
- [x] I have verified that I am following the CDP4-SDK [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/CDP4-WebServices-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description

- [x] Added BeforeDelete method in IterationSetupSideEffect, where validate IterationSetup.isDeleted an IterationSetup.FrozenOn.
- [x] Added test that check solution.

### CODE REVIEW

File: CDP4Orm/Dao/Supplemental/IterationSetupDao.cs [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ﻿// --------------------------------------------------------------------------------------------------------------------
 // <copyright file="IterationSetupDao.cs" company="RHEA System S.A.">
-//   Copyright (c) 2016 RHEA System S.A.
+//   Copyright (c) 2016-2020 RHEA System S.A.
 // </copyright>
 // --------------------------------------------------------------------------------------------------------------------
```
Comment (nathanatrhea) [inline CDP4Orm/Dao/Supplemental/IterationSetupDao.cs:5]: I think we have a newest copyright string

File: CDP4WebServices.API.Tests/SideEffects/IterationSetupSideEffectTestFixture.cs [hunk 2]
```diff
@@ -1,6 +1,6 @@
 ﻿// --------------------------------------------------------------------------------------------------------------------
 // <copyright file="IterationSetupSideEffectTestFixture.cs" company="RHEA System S.A.">
-//   Copyright (c) 2016 RHEA System S.A.
+//   Copyright (c) 2016-2020 RHEA System S.A.
 // </copyright>
```
Comment (nathanatrhea) [inline CDP4WebServices.API.Tests/SideEffects/IterationSetupSideEffectTestFixture.cs:4]: I think we have a newest copyright string

File: CDP4WebServices.API/Services/Operations/SideEffects/Implementation/IterationSetupSideEffect.cs [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ﻿// --------------------------------------------------------------------------------------------------------------------
 // <copyright file="IterationSetupSideEffect.cs" company="RHEA System S.A.">
-//   Copyright (c) 2016-2018 RHEA System S.A.
+//   Copyright (c) 2016-2020 RHEA System S.A.
 // </copyright>
```
Comment (nathanatrhea) [inline CDP4WebServices.API/Services/Operations/SideEffects/Implementation/IterationSetupSideEffect.cs:4]: I think we have a newest copyright string

File: CDP4WebServices.API/Services/Operations/SideEffects/Implementation/IterationSetupSideEffect.cs [hunk 4]
```diff
@@ -203,11 +222,26 @@ public override void AfterCreate(IterationSetup thing, Thing container, Iteratio
...
             {
                 throw new InvalidOperationException("It is not possible to delete the current iteration.");
             }
 
+            if (iterationSetup.IsDeleted != false)
+            {
+                return;
+            }
+
+            iterationSetup.IsDeleted = true;
```
Comment (samatrhea) [inline CDP4WebServices.API/Services/Operations/SideEffects/Implementation/IterationSetupSideEffect.cs:243]: we reach line 243 when the `iterationSetup.IsDeleted = true`, why do we need to set it and then persist that in the dbase?

File: CDP4Orm/Dao/Supplemental/IterationSetupDao.cs [hunk 5]
```diff
@@ -106,5 +125,32 @@ public IEnumerable<IterationSetup> ReadByEngineeringModelSetup(NpgsqlTransaction
...
+        /// </param>
+        /// <returns>
+        /// True if the concept was deleted.
+        /// </returns>
+        /// <remarks>
+        /// IterationSetups cannot be deleted at all.
+        /// </remarks>>
+        public override bool BeforeDelete(NpgsqlTransaction transaction, string partition, Guid iid, out bool isHandled)
+        {
+            isHandled = true;
```
Comment (samatrhea) [inline CDP4Orm/Dao/Supplemental/IterationSetupDao.cs:152]: why do we have this override?

File: CDP4WebServices.API/Services/Operations/SideEffects/Implementation/IterationSetupSideEffect.cs (hunk 4 above)
Comment (kamilatrhea) [reply CDP4WebServices.API/Services/Operations/SideEffects/Implementation/IterationSetupSideEffect.cs:243]: Deleting the contents of an iteration means removing the Iteration and all the objects it contains from the data store, but we have to remain descriptive data in IterationSetup for other objects, so we cannot delete iterationSetup, but only mark it like a deleted. This approach prevent any wrong read of the associated Iteration data (on any level within the Iteration) and prevent showing deleted iterations when requesting.

File: CDP4Orm/Dao/Supplemental/IterationSetupDao.cs (hunk 5 above)
Comment (alexatrhea) [reply CDP4Orm/Dao/Supplemental/IterationSetupDao.cs:152]: I think this is the only way to jump out of a delete prematurely.

Comment (lxatrhea) [reply CDP4Orm/Dao/Supplemental/IterationSetupDao.cs:152]: I think @alexatrhea is right.

**Pull request merged**